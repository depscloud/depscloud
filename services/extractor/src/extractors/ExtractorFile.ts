import cheerio = require("cheerio");
import YAML = require("js-yaml");
import TOML = require("toml");

export default class ExtractorFile {
    private readonly body: string;

    constructor(body: string) {
        this.body = body;
    }

    public raw(): string {
        return this.body;
    }

    public json(): any {
        return JSON.parse(this.body);
    }

    public toml(): any {
        return TOML.parse(this.body);
    }

    public yaml(): any {
        return YAML.safeLoad(this.body);
    }

    public xml(): cheerio.Cheerio {
        return cheerio.load(this.body).root();
    }
}
