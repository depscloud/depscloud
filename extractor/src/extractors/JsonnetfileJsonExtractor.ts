import {Dependency, DependencyManagementFile} from "@depscloud/api/v1alpha/deps";
import MatchConfig from "../matcher/MatchConfig";
import Extractor from "./Extractor";
import ExtractorFile from "./ExtractorFile";
import Languages from "./Languages";

const extract = (dependency: any):Dependency => {
    const { source, version } = dependency;
    const { git } = (source || {});
    const { subdir, remote } = (git || {});

    return {
        name: remote,
        versionConstraint: version,
        scopes: subdir ? [ subdir ] : [],
        organization: "", // deprecated
        module: "", // deprecated
    };
}

export default class JsonnetfileJsonExtractor implements Extractor {
    matchConfig(): MatchConfig {
        return {
            includes: [
                "**/jsonnetfile.json",
            ],
            excludes: [],
        };
    }
    requires(): string[] {
        return [ "jsonnetfile.json" ];
    }
    public async extract(url: string, files: { [key: string]: ExtractorFile }): Promise<DependencyManagementFile> {
        const {
            version,
            dependencies,
        } = files["jsonnetfile.json"].json();

        const deps = dependencies.map(dependency => extract(dependency));


        return {
            language: Languages.JSONNET,
            system:"jsonnet-bundler",
            dependencies: deps,
            name: url,
            organization:"",
            module:"", //deprecated
            sourceUrl:"",
            version
        }
    }
}
