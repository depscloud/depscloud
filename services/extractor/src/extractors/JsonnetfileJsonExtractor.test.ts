import {readFile} from "fs";
import {promisify} from "util";
import ExtractorFile from "./ExtractorFile";
import JsonnetfileJsonExtractor from "./JsonnetfileJsonExtractor";

const readFileAsync = promisify(readFile);

describe("JsonnetfileJsonExtractor", () => {
    test("fullParse", async () => {
        const jsonPath = require.resolve("./testdata/jsonnetfile.json");
        const buffer = await readFileAsync(jsonPath);
        
        const content = buffer.toString();

        const parser = new JsonnetfileJsonExtractor();

        const actual = await parser.extract("", { "jsonnetfile.json": new ExtractorFile(content) });

        expect(actual).toMatchSnapshot();
    });
});
