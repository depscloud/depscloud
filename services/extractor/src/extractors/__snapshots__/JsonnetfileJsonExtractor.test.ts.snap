// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonnetfileJsonExtractor fullParse 1`] = `
Object {
  "dependencies": Array [
    Object {
      "name": "https://github.com/grafana/grafonnet-lib.git",
      "scopes": Array [
        "grafonnet",
      ],
      "versionConstraint": "master",
    },
    Object {
      "name": "https://github.com/grafana/jsonnet-libs.git",
      "scopes": Array [
        "grafana-builder",
      ],
      "versionConstraint": "master",
    },
  ],
  "language": "jsonnet",
  "name": "",
  "sourceUrl": "",
  "system": "jsonnet-bundler",
  "version": 1,
}
`;
